src/pulsar_producer.erl

```diff
@@ -49,233 +90,579 @@ callback_mode() -> [state_functions].
+    NewQ = replayq:append(Q, QItems),
+    State#{replayq := NewQ}.
+
+maybe_send_to_pulsar(State0) ->
+    #{ batch_size := BatchSize
+     , sequence_id := SequenceId
+     , requests := Requests0
+     , replayq := Q
+     , opts := ProducerOpts
+     } = State0,
+    case replayq:count(Q) =:= 0 of
+        true ->
+            State0;
+        false ->
+            MaxBatchBytes = maps:get(max_batch_bytes, ProducerOpts, ?DEFAULT_MAX_BATCH_BYTES),
+            {NewQ, QAckRef, Items} = replayq:pop(Q, #{ count_limit => BatchSize
+                                                     , bytes_limit => MaxBatchBytes
+                                                     }),
+            State1 = State0#{replayq := NewQ},
+            RetentionPeriod = maps:get(retention_period, ProducerOpts, infinity),
+            Now = now_ts(),
+            {Froms, Messages} =
+                lists:foldr(
+                  fun(?Q_ITEM(From, Timestamp, Msgs), {Froms, AccMsgs}) ->
+                    case is_batch_expired(Timestamp, RetentionPeriod, Now) of
+                      true -> {Froms, AccMsgs};
+                      false -> {[From | Froms], [{Timestamp, Msgs} | AccMsgs]}
+                    end
+                  end,
+                  {[], []},
+                  Items),
+            case Messages of
+                [] ->
+                    %% all expired, immediately ack replayq batch and continue
+                    ok = replayq:ack(Q, QAckRef),
+                    maybe_send_to_pulsar(State1);
+                [_ | _] ->
+                    send_batch_payload([Msg || {_Timestamp, Msgs} <- Messages, Msg <- Msgs],
+                                       SequenceId, State0),
+                    Requests = Requests0#{SequenceId => {QAckRef, Froms, Messages}},
+                    State2 = State1#{requests := Requests},
+                    State = next_sequence_id(State2),
+                    maybe_send_to_pulsar(State)
+            end
+    end.
 
-next_request_id(State = #state{request_id = ?MAX_QUE_ID}) ->
-    State#state{request_id = 1};
-next_request_id(State = #state{request_id = RequestId}) ->
-    State#state{request_id = RequestId+1}.
+collect_send_requests(Acc, Limit) ->
+    Count = 0,
+    do_collect_send_requests(Acc, Count, Limit).
+
+do_collect_send_requests(Acc, Count, Limit) when Count >= Limit ->
+    lists:reverse(Acc);
+do_collect_send_requests(Acc, Count, Limit) ->
+    receive
+        ?SEND_REQ(_, _) = Req ->
+            do_collect_send_requests([Req | Acc], Count + 1, Limit)
+    after
+        0 ->
+            lists:reverse(Acc)
+    end.
 
-next_sequence_id(State = #state{sequence_id = ?MAX_SEQ_ID}) ->
-    State#state{sequence_id = 1};
-next_sequence_id(State = #state{sequence_id = SequenceId}) ->
-    State#state{sequence_id = SequenceId+1}.
+try_close_socket(#{sock := undefined}) ->
+    ok;
+try_close_socket(#{sock := Sock, opts := Opts}) ->
+    catch pulsar_socket:close(Sock, Opts),
```

Comment:
not sure if later version Erlang compiler optimised `catch` to avoid retrieving stacktrace in case the result is discarded, the old school recommendation is: avoid using `catch` because it generates stacktrace which is unnecessary expensive.